content/news/2022-04-12-bevy-0.7/index.md

```diff
@@ -0,0 +1,654 @@
+
+<div class="release-feature-authors">authors: @MrGVSV</div>
+
+Rust encourages the use of the [newtype pattern](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) when expanding a type with new functionality or meaning. This is also a useful tool in Bevy:
+
+```rust
+#[derive(Component)]
+struct Items(Vec<Item>);
+
+fn give_sword(mut query: Query<&mut Items>) { 
+    for mut items in query.iter_mut() {
+        items.0.push(Item::new("Flaming Poisoning Raging Sword of Doom"));
+    }
+}
+```
+
+This works just fine, but that `0` at the end of `items.0` sticks out like a sore thumb. Many of us in the Bevy Org think `.0` has no place in public apis. But the newtype pattern is still useful! Ideally, Rust would provide a way to express that `Items` is a new type, while transparently provided access to the `Vec<Item>` stored within. There are designs being discussed by the Rust team, but we don't want to wait for nice things!
+
+Fortunately, the Deref / DerefMut traits in std provide the behavior we want. Users can already manually implement these traits, but for such a common pattern, we decided that providing our own trait derives was worth it. In **Bevy 0.7**, you can now derive Deref and DerefMut, enabling much nicer public apis:
+
+```rust
+#[derive(Component, Deref, DerefMut)]
+struct Items(Vec<Item>);
+
+fn give_sword(mut query: Query<&mut Items>) { 
+    for mut items in query.iter_mut() {
+        // No more .0!
+        items.push(Item::new("Flaming Poisoning Raging Sword of Doom"));
+    }
+}
+```
+
+Astute `std` doc readers might notice that the Rust team [recommends only using `Deref`/`DerefMut` for smart pointers, to avoid confusion](https://doc.rust-lang.org/std/ops/trait.Deref.html). Components like `Items` _are not_ smart pointers. We choose to ignore this advice, as this pattern works, is already widely used in the Rust ecosystem, and Good UX Comes First.
+
+## WorldQuery Derives
+
+<div class="release-feature-authors">authors: @mvlabat</div>
+
+Sometimes when building Bevy Apps you might find yourself repeating the same sets of components over and over in your queries:
+
+```rust
+fn move_players(mut players: Query<(&mut Name, &mut Transform, &mut Item)>) {
+    for (name, transform, item) in players.iter_mut() {
+    }
+}
+
+fn despawn_players(mut players: Query<(Entity, &mut Name, &mut Transform, &mut Item)>) {
+    for (entity, name, transform, item) in players.iter_mut() {
+    }
+}
+```
+
+Maybe you've gotten tired of typing the same components over and over. In **Bevy 0.7**, you can now easily create your own custom `WorldQuery` trait implementations with the `WorldQuery` derive:
+
+```rust
+#[derive(WorldQuery)]
+struct PlayerQuery<'w> {
+    name: &'w mut Name,
+    transform: &'w mut Transform,
+    items: &'w mut Items,
+}
+
+fn move_players(mut players: Query<PlayerQuery>) {
+    for player in players.iter_mut() {
+    }
+}
+
+fn despawn_players(mut players: Query<(Entity, PlayerQuery)>) {
```

Comment:
It looks like the `PlayerQuery` can represent a commonly-queried set of components, and you can arbitrarily extend it with other components in the query. That is super cool!

Reply:
Agreed. I dig it!